syntax = "proto3";

package todo;

message Todo {
    int32 id = 1;
    string title = 2;
    string description = 3;
    bool completed = 4;
}

message CreateTodoRequest {
    string title = 1;
    string description = 2;
}

message CreateTodoResponse {
    int32 id = 1;
    string title = 2;
    string description = 3;
    bool completed = 4;
}

message GetTodoRequest {
    int32 id = 1;
}

message GetTodoResponse {
    int32 id = 1;
    string title = 2;
    string description = 3;
    bool completed = 4;
}

message UpdateTodoRequest {
    int32 id = 1;
    optional string title = 2;
    optional string description = 3;
    optional bool completed = 4;
}

message UpdateTodoResponse {
    int32 id = 1;
    string title = 2;
    string description = 3;
    bool completed = 4;
}

message DeleteTodoRequest {
    int32 id = 1;
}

message DeleteTodoResponse {
    bool success = 1;
}

message GetAllTodoRequest {}

message GetAllTodoResponse {
    repeated Todo todos = 1;
}

service TodoService {
    rpc CreateTodo(CreateTodoRequest) returns (CreateTodoResponse);
    rpc GetTodo(GetTodoRequest) returns (GetTodoResponse);
    rpc UpdateTodo(UpdateTodoRequest) returns (UpdateTodoResponse);
    rpc DeleteTodo(DeleteTodoRequest) returns (DeleteTodoResponse);
    rpc GetAllTodos(GetAllTodoRequest) returns (GetAllTodoResponse);
}